File: service/pyShobuService/src/logic/shobu/gameoverChecker.py


import logic.shobu.move as M
import logic.shobu.helpers as h
import logic.shobu.token as t
import logic.shobu.board as B
from exception.ExceptionToReturn import ExceptionToReturn

def checkForWin( board ):
    for n in range( 4 ):
        subboard = B.getSubboard( board, n )
        curBlack, curWhite = _countTokenTypeOnSubboard( subboard )
        if curBlack == 0:
            return t.SIDE_WHITE
        if curWhite == 0:
            return t.SIDE_BLACK
    return None

def _countTokenTypeOnSubboard( subboard ):
    blackCount, whiteCount = 0, 0
    for spot in h.everySpot():
        foundToken = B.subboardGetToken( subboard, spot )
        if foundToken == t.BLACK:
            blackCount += 1
        elif foundToken == t.WHITE:
            whiteCount += 1
    return blackCount, whiteCount

File: client/shobu_client/src/gameLogic/gameoverChecker.js
import { getSubboard, subboardGetToken } from "./board";
import { token } from "./token";

export function checkForWin( board ){
    for( let i=0; i<4; i++ ){
        let subboard = getSubboard( board, i );
        let [ blackCount, whiteCount ] = countTokenTypeOnSubboard( subboard );
        if( blackCount == 0 ){
            return token.WHITE;
        }
        if( whiteCount == 0 ){
            return token.BLACK;
        }
    }
    return null
}

function countTokenTypeOnSubboard( subboard ){
    let blackCount = 0;
    let whiteCount = 0;
    for( let iy=0; iy<4; iy++ ){
        for( let ix=0; ix<4; ix++ ){
            let foundToken = subboardGetToken( subboard, [ix,iy] ).type;
            if( foundToken == token.BLACK ){
                blackCount += 1;
            }
            else if( foundToken == token.WHITE ){
                whiteCount += 1;
            }
        }
    }
    return [ blackCount, whiteCount ];
}